src/crimson/os/seastore/segment_manager/zns.cc

```diff
@@ -590,17 +651,78 @@ Segment::write_ertr::future<> ZNSSegment::write(
   seastore_off_t offset, ceph::bufferlist bl)
 {
   LOG_PREFIX(ZNSSegment::write);
-  if (offset < write_pointer || offset % manager.metadata.block_size != 0) {
-    ERROR("invalid segment write on segment {} to offset {}",
-      id,
-      offset);
+  if (offset != write_pointer || offset % manager.metadata.block_size != 0) {
+    ERROR("Segment offset and zone write pointer mismatch. "
+          "segment {} segment-offset {} write pointer {}",
+          id, offset, write_pointer);
     return crimson::ct_error::invarg::make();
   }
-  if (offset + bl.length() > manager.metadata.segment_size)
+  if (offset + bl.length() > manager.metadata.segment_capacity) {
     return crimson::ct_error::enospc::make();
+  }
   
   write_pointer = offset + bl.length();
   return manager.segment_write(paddr_t::make_seg_paddr(id, offset), bl);
 }
 
+Segment::write_ertr::future<> ZNSSegment::write_padding_bytes(
+  size_t padding_bytes)
+{
+  LOG_PREFIX(ZNSSegment::write_padding_bytes);
+  DEBUG("Writing {} padding bytes to segment {} at wp {}",
+        padding_bytes, id, write_pointer);
+
+  return crimson::repeat([FNAME, padding_bytes, this] () mutable {
+    size_t bufsize = 0;
+    if (padding_bytes >= MAX_PADDING_SIZE) {
+      bufsize = MAX_PADDING_SIZE;
+    } else {
+      bufsize = padding_bytes;
+    }
+
+    padding_bytes -= bufsize;
+    bufferptr bp(ceph::buffer::create_page_aligned(bufsize));
+    bp.zero();
+    bufferlist padd_bl;
+    padd_bl.append(bp);
+    return write(write_pointer, padd_bl).safe_then([FNAME, padding_bytes, this]() {
+      if (padding_bytes == 0) {
+        return write_ertr::make_ready_future<seastar::stop_iteration>(seastar::stop_iteration::yes);
+      } else {
+        return write_ertr::make_ready_future<seastar::stop_iteration>(seastar::stop_iteration::no);
+      }
+    });
+  }).safe_then([]() {
+    return write_ertr::now();
+  }).handle_error(
+   write_ertr::pass_further{},
+   crimson::ct_error::assert_all{
+     "crimson::repeat() error in ZNSSegment::write_padding_bytes"
+   });
```

Comment:
Seems unnecessary that can be dropped.
```c++
  }).safe_then([]() {
    return write_ertr::now();
  }).handle_error(
   write_ertr::pass_further{},
   crimson::ct_error::assert_all{
     "crimson::repeat() error in ZNSSegment::write_padding_bytes"
   });
```

Reply:
Shouldn't there be an error handler here, I thought the error messages might help in debugging.

Reply:
There is no error message with this code, all possible errors are forwarded by `write_ertr::pass_further{}` in this case.

It is only necessary to handle error if `ZNSSegment::write()` returns more errors than `ZNSSegment::write_padding_bytes()`.